### ISSUE TITLE
Loading snippets from unpacked tables leads to missing snippets

### ISSUE THREAD

**Issue by jtws** [L3MON4D3/LuaSnip#718]
I use the lua lazy loader on a specific directory where I keep all LuaSnip lua snippets. One file per filetype. I separate snippets to tables and unpack them in the return call. This leads to some unpacked snippets not being loaded, often only the first snippet in an unpacked table is loaded.  I'm running off master and debug logging shows no errors, only snippet load counts which are missing 90% of snippets.

## Bug/what I'm doing wrong?

If I put snippets in tables like `snip.xxx`, `snip.yyy`, `snip.zzz` to organise them and then `return { unpack(snip.xxx), unpack(snip.yyy), unpack.snip(zzz) }, { ... same for autosnips }` all snippets in some tables are loaded and in some tables only the first snippet gets loaded. I'm running off master with debug logging and when I reload snippets there are no errors, only log messages like "<file> loaded 6 autosnippets" when that file defined 30 autosnippets.

If I copy-paste verbatim the same snippets directly into the `return {}, {}` tables, then all snippets get loaded. So it's not an issue with snippet syntax breaking parsing in certain tables or something like that. The snippets are fine.

I'm not a Lua expert and can only guess the unpack calls don't finish before return. How that's possible I don't know. I tried lazy loading and loading directly with the same results.

What I didn't test was just recursively flattening tables to two intermediary tables and then returning those. Maybe that's the way forward? But I still want to understand why unpacking in return call leads to one snippet being loaded from the unpacked table and the rest are not loaded. Clearly the unpacking happened before return but for some reason only one snippet gets loaded.

### Why this is an issue

Returning two massive tables (snippets, autosnippets) makes navigation hard (can't use an outliner at all) and damn Stylua likes to add an indent to the second returned table, leading to different indentation for regular and auto snippets, making it extra ugly.

By organizing snippets per table by their type/usecase I can quickly jump to a specific table with an outliner and move with text objects easily. I also have LSP context in statusbar so I know what snippet table I happen to be editing and don't get lost.

As a workaround I had to add several levels of section dividers in the returned tables to have some kind of organisation but it's still ugly and code is unpleasant to work with. I like clean code.

**Comment by L3MON4D3** [comment]
Oh, I think this is just lua being lua, when unpacking into a function call (I think it's the same for constructing tables) all unpacks, except the last one, only pass one item.

I agree that this is an issue, but we also have a fix for that, see 8145f4b, or [my config](https://github.com/L3MON4D3/Dotfiles/blob/362701be7711049196a92bc12d40a8a38b7dba03/.config/nvim/lua/plugins/luasnip/init.lua#L97-L106)

**Comment by jtws** [comment]
> Oh, I think this is just lua being lua, when unpacking into a function call (I think it's the same for constructing tables) all unpacks, except the last one, only pass one item.
> 
> I agree that this is an issue, but we also have a fix for that, see [8145f4b](https://github.com/L3MON4D3/LuaSnip/commit/8145f4b8f8f2e1c98a1f9914228f910ca36a5d69), or [my config](https://github.com/L3MON4D3/Dotfiles/blob/362701be7711049196a92bc12d40a8a38b7dba03/.config/nvim/lua/plugins/luasnip/init.lua#L97-L106)

Lua... I tried the loading tables and they're working for s(nippets) and as(nippets), but the parse function in my custom `snip_env` was not available in my snippet files. I tried calling `ls.setup_snip_env()` manually but that resulted in another error about inserting a nil value IIRC. After I removed the manual call the global s and as (what I call the snip and autosnip loader funcs) started working and logs show both snips and autosnip are being loaded from Lua files.

Here's my `ls.setup(), I copied it mostly from your example. In case its useful for someone. The broken parser turned out to be wrong/outdated documentation, as explained at the end.

<details><summary>ls.setup()</summary>

```lua
ls.setup {
    snip_env = {
        -- Snippets defined with this s are automatically loaded
        s = function(...)
            local snip = ls.s(...)
            -- we can't just access the global `ls_file_snippets`, since it will be
            -- resolved in the environment of the scope in which it was defined.
            table.insert(getfenv(2).ls_file_snippets, snip)
        end,
        --  As are snippets parsed with this.
        parse = function(...)
            local snip = ls.parser.parse_snippet(...)
            table.insert(getfenv(2).ls_file_snippets, snip)
        end,
        -- Autosnippets defined with this are autoloaded
        as = function(...)
            local snip = ls.s(...)
            -- we can't just access the global `ls_file_snippets`, since it will be
            -- resolved in the environment of the scope in which it was defined.
            table.insert(getfenv(2).ls_file_autosnippets, snip)
        end,
        --  As are autosnippets parsed with this.
        aparse = function(...)
            --[[ local snip = ls.parser.parse(...) ]]
            local snip = ls.parser.parse_snippet(...)
            table.insert(getfenv(2).ls_file_autosnippets, snip)
        end,
        -- Rest of standard nodes with LuaSnip authors customisations I don't understand but I'll use his config.
        sn = ls.sn,
        t = ls.t,
        i = ls.i,
        f = function(func, argnodes, ...)
            return ls.f(
                function(args, imm_parent, user_args) return func(args, imm_parent.snippet, user_args) end,
                argnodes,
                ...
            )
        end,
        -- override to enable restore_cursor.
        c = function(pos, nodes, opts)
            opts = opts or {}
            opts.restore_cursor = true
            return ls.c(pos, nodes, opts)
        end,
        d = function(pos, func, argnodes, ...)
            return ls.d(
                pos,
                function(args, imm_parent, old_state, ...) return func(args, imm_parent.snippet, old_state, ...) end,
                argnodes,
                ...
            )
        end,
        isn = require("luasnip.nodes.snippet").ISN,
        l = require("luasnip.extras").lambda,
        dl = require("luasnip.extras").dynamic_lambda,
        rep = require("luasnip.extras").rep,
        r = ls.restore_node,
        p = require("luasnip.extras").partial,
        types = require("luasnip.util.types"),
        events = require("luasnip.util.events"),
        util = require("luasnip.util.util"),
        fmt = require("luasnip.extras.fmt").fmt,
        fmta = require("luasnip.extras.fmt").fmta,
        ls = ls,
        ins_generate = function(nodes)
            return setmetatable(nodes or {}, {
                __index = function(table, key)
                    local indx = tonumber(key)
                    if indx then
                        local val = ls.i(indx)
                        rawset(table, key, val)
                        return val
                    end
                end,
            })
        end,
        n = require("luasnip.extras").nonempty,
        m = require("luasnip.extras").match,
        ai = require("luasnip.nodes.absolute_indexer"),
        postfix = require("luasnip.extras.postfix").postfix,
        conds = require("luasnip.extras.expand_conditions"),
    },
    -- ft_func = ft_functions.from_filetype,
	-- fn(bufnr) -> string[] (filetypes).
	-- load_ft_func = ft_functions.from_filetype_load,
    ft_func = require("luasnip.extras.filetype_functions").from_pos_or_filetype,
    load_ft_func = require("luasnip.extras.filetype_functions").extend_load_ft {
        markdown = { "lua", "json", "python", "sh", "vim", "go", "zsh" },
        zsh = { "sh" },
    },
    -- Update more often, :h events for more info.
    region_check_events = "CursorHold,InsertLeave",
    delete_check_events = "TextChanged,InsertEnter",
    update_events = "TextChanged,TextChangedI",
    enable_autosnippets = true,
    history = true,
    --delete_check_events = "TextChanged",
    -- Show node indicators
    ext_opts = {
        [types.choiceNode] = {
            active = {
                virt_text = { { "●", "Type" } }, -- yellow
            },
        },
        [types.insertNode] = {
            active = {
                virt_text = { { "●", "Include" } }, -- blue
            },
        },
    },
    -- Required for SELECT vars to be populated. Select in visual ans hit tab
    -- to populate vars, then select snippet. Cleared after snippet exits.
    store_selection_keys = "<Tab>",
}
```
</details>

However, the [DOC.md Lua loading section](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#lua) has code that doesn't work. It uses

```lua
                parse = function(...)
			local snip = ls.parser.parse(...)
			table.insert(getfenv(2).ls_file_snippets, snip)
		end,
```

and `ls.parser.parse() doesn't exist. I had to replace it with:

```lua
        parse = function(...)
            local snip = ls.parser.parse_snippet(...)
            table.insert(getfenv(2).ls_file_snippets, snip)
        end,
```

So `ls.parser.parse_snippet(...)` instead of `ls.parser.parse(...)`.

After that change both LuaSnip s and as (autosnip) and parse and aparse (autosnip) loaders started working properly.

I read through this part of the doc probably four times but ended up thinking returning two tables was the new way together with using lazy)_oad(), I thought lazy_load() was an alternative to specifying a custom snip env from the wording. But now things are clearer and everything's working. Thanks again!

**Comment by L3MON4D3** [comment]
Glad it works! 
Very sorry about the outdated documentation! :/
I'll update that asap
